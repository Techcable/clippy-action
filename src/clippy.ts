/*
 * üêª‚Äç‚ùÑÔ∏èüì¶ clippy-action: GitHub action to run Clippy, an up-to-date and modern version of actions-rs/clippy
 * Copyright 2023 Noel Towa <cutie@floofy.dev>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { type AnnotationProperties, info, error, warning, debug, startGroup, endGroup, summary } from '@actions/core';
import { assertIsError, hasOwnProperty } from '@noelware/utils';
import { type ExecOptions, exec } from '@actions/exec';
import type { Inputs } from './inputs';
import { arch, os } from './os-info';

// We only use a renderer so we can mock it in tests
export interface Renderer {
    warning(message: string, properties: AnnotationProperties): void;
    error(message: string, properties: AnnotationProperties): void;
    info(message: string): void;
}

class DefaultRenderer implements Renderer {
    annotations: (AnnotationProperties & { level: 'error' | 'warning'; rendered: string })[] = [];

    warning(message: string, properties: AnnotationProperties) {
        this.annotations.push({ ...properties, level: 'warning', rendered: message });
        warning(message, properties);
    }

    error(message: string, properties: AnnotationProperties) {
        this.annotations.push({ ...properties, level: 'error', rendered: message });
        error(message, properties);
    }

    info(message: string) {
        info(message);
    }
}

export const kDefaultRenderer = new DefaultRenderer();

export const getClippyOutput = async (
    inputs: Inputs,
    cargoPath: string
): Promise<[exitCode: number, data: string[]]> => {
    startGroup('Executing `cargo clippy`...');

    const args = ['clippy', '--message-format', 'json'];
    if (inputs['all-features']) {
        args.push('--all-features');
    }

    if (inputs.forbid.length) {
        args.push(...inputs.forbid.map((forbid) => `-F${forbid}`));
    }

    if (inputs.deny.length) {
        args.push(...inputs.deny.map((deny) => `-D${deny}`));
    }

    if (inputs.warn.length) {
        args.push(...inputs.warn.map((allowed) => `-W${allowed}`));
    }

    if (inputs.allow.length) {
        args.push(...inputs.allow.map((allowed) => `-A${allowed}`));
    }

    const data: string[] = [];
    const execOptions: ExecOptions = {
        ignoreReturnCode: true,
        failOnStdErr: false,
        listeners: {
            stdline(piece) {
                try {
                    JSON.parse(piece);
                    data.push(piece);
                } catch (e) {
                    assertIsError(e);

                    // we don't care if invalid json
                    // was passed, we won't include it
                    // anyway.
                }
            }
        }
    };

    if (inputs['working-directory'] !== undefined) {
        execOptions.cwd = inputs['working-directory'];
    }

    const exitCode = await exec(cargoPath, args, execOptions);

    endGroup();
    return [exitCode, data];
};

export const renderMessages = async (pieces: string[], renderer: Renderer = kDefaultRenderer) => {
    for (const piece of pieces) {
        const data = JSON.parse(piece);
        if (!hasOwnProperty(data, 'reason')) {
            debug(`Skipping payload due to no 'reason' output\n${piece}`);
        }

        if (!hasOwnProperty(data, 'message')) {
            debug(`Skipping payload due to no 'message' output\n${piece}`);
            continue;
        }

        if (data.reason !== 'compiler-message') {
            debug(`Skipping payload due to 'reason' !== 'compiler-message'\n${piece}`);
            continue;
        }

        if (!hasOwnProperty(data, 'target')) {
            debug(`Skipping payload due to no 'target' output\n${piece}`);
            continue;
        }

        const message = data.message;
        let type: 'info' | 'warning' | 'error';
        let method: any;

        switch (message.level) {
            case 'help':
            case 'note':
                type = 'info';
                method = renderer.info;

                break;

            case 'warning':
                type = 'warning';
                method = renderer.warning;

                break;

            case 'error':
                type = 'error';
                method = renderer.error;

                break;

            case 'error: internal compiler error':
            default:
                error(
                    'Received a internal compiler error OR an unknown message type, view this in debug mode to view the payload'
                );

                error(piece);
                process.exit(1);
        }

        // Don't log "error: aborting due to" messages
        if (message.rendered.includes('error: aborting due to previous error')) {
            continue;
        }

        const primarySpan = message.spans.find((span: any) => hasOwnProperty(span, 'is_primary') && span.is_primary);
        if (primarySpan !== undefined) {
            const args =
                type === 'info'
                    ? [message.rendered]
                    : [
                          message.rendered,

                          // We shouldn't include file as tests will be flaky
                          // in CI, so it'll only show up if it was
                          // actually ran.
                          process.env.VITEST === 'true'
                              ? ({
                                    startColumn: primarySpan.column_start,
                                    endColumn: primarySpan.column_end,
                                    startLine: primarySpan.line_start,
                                    endLine: primarySpan.line_end,
                                    title: message.message
                                } satisfies AnnotationProperties)
                              : ({
                                    startColumn: primarySpan.column_start,
                                    endColumn: primarySpan.column_end,
                                    startLine: primarySpan.line_start,
                                    endLine: primarySpan.line_end,
                                    title: message.message,
                                    file: data.target.src_path
                                } satisfies AnnotationProperties)
                      ];

            method.apply(renderer, args);
        } else {
            method(message.rendered);
        }
    }
};
