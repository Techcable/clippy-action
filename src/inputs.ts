/*
 * üêª‚Äç‚ùÑÔ∏èüì¶ clippy-action: GitHub action to run Clippy, an up-to-date and modern version of actions-rs/clippy
 * Copyright 2023 Noel Towa <cutie@floofy.dev>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { error, getInput, type InputOptions } from '@actions/core';
import z from 'zod';

const inputSchema = z.object({
    'all-features': z.boolean().default(false),
    forbid: z.array(z.string()).default([]),
    allow: z.array(z.string()).default([]),
    deny: z.array(z.string()).default([]),
    args: z.array(z.string()).default([]),
    warn: z.array(z.string()).default([])
});

export type Inputs = z.infer<typeof inputSchema>;

const truthy = new Set(['true', 'True', 'TRUE']);
const falsy = new Set(['false', 'False', 'FALSE']);

const getBooleanInput = (name: string, options: InputOptions) => {
    const value = getInput(name, options);
    if (!value) return false;
    if (truthy.has(value)) return true;
    if (falsy.has(value)) return false;

    throw new Error(
        `Value of key [${name}] didn't meet the Yaml 1.2 "Core Schema" specification (received: [${value}])`
    );
};

export const getInputs = (): Inputs | null => {
    const allFeatures = getBooleanInput('all-features', { trimWhitespace: true });
    const allow = getInput('allow', { trimWhitespace: true })
        .split(',')
        .map((s) => s.trim());

    const deny = getInput('deny', { trimWhitespace: true })
        .split(',')
        .map((s) => s.trim());

    const forbid = getInput('forbid', { trimWhitespace: true })
        .split(',')
        .map((s) => s.trim());

    const warn = getInput('forbid', { trimWhitespace: true })
        .split(',')
        .map((s) => s.trim());

    const args = getInput('args', { trimWhitespace: true })
        .split(',')
        .map((s) => s.trim());

    if (args.some((s) => s.startsWith('-W') || s.startsWith('--warning'))) {
        error('To append new warning lints, use the `warn` action input.');
        return null;
    }

    if (args.some((s) => s.startsWith('-D') || s.startsWith('--deny'))) {
        error('To append new deny lints, use the `deny` action input.');
        return null;
    }

    if (args.some((s) => s.startsWith('-F') || s.startsWith('--forbid'))) {
        error('To append new forbidden lints, use the `forbid` action input.');
        return null;
    }

    if (args.some((s) => s.startsWith('-A') || s.startsWith('--allow'))) {
        error('To append new allowed lints, use the `allow` action input.');
        return null;
    }

    return {
        'all-features': allFeatures,
        forbid,
        allow,
        deny,
        warn,
        args
    };
};
