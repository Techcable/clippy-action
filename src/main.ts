/*
 * 🐻‍❄️📦 clippy-action: GitHub action to run Clippy, an up-to-date and modern version of actions-rs/clippy
 * Copyright 2023 Noel Towa <cutie@floofy.dev>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { endGroup, error, info, startGroup } from '@actions/core';
import { getOctokit, context } from '@actions/github';
import { getInputs } from './inputs';
import * as clippy from './clippy';
import { which } from '@actions/io';

async function main() {
    const inputs = await getInputs();
    if (inputs === null) {
        process.exit(1);
    }

    const client = getOctokit(inputs['github-token']);

    startGroup('Checking if `cargo` exists...');
    let cargoPath: string;

    try {
        cargoPath = await which('cargo', true);
        info(`Cargo exists in path [${cargoPath}]`);
    } catch (e) {
        error("Tool doesn't exist, please add a valid Rust toolchain.");
        process.exit(1);
    } finally {
        endGroup();
    }

    // Create a check
    const {
        data: { id }
    } = await client.request('POST /repos/{owner}/{repo}/check-runs', {
        owner: context.repo.owner,
        repo: context.repo.repo,

        name: 'Clippy',
        head_sha: context.payload.pull_request !== undefined ? context.payload.pull_request.head.sha : context.sha,
        status: 'in_progress'
    });

    const [exitCode, pieces] = await clippy.getClippyOutput(inputs, cargoPath);
    await clippy.renderMessages(pieces);

    const annotations = clippy.kDefaultRenderer.annotations;
    await client.request('PATCH /repos/{owner}/{repo}/check-runs/{check_run_id}', {
        check_run_id: id,
        owner: context.repo.owner,
        repo: context.repo.repo,
        conclusion: exitCode === 0 ? 'success' : 'failure',
        output:
            exitCode === 0
                ? {
                      title: 'Clippy Result',
                      summary: 'Clippy ran successfully!',
                      annotations: annotations.map((anno) => ({
                          annotation_level: anno.level === 'error' ? ('failure' as const) : ('warning' as const),
                          path: anno.file!,
                          start_line: anno.startLine!,
                          end_line: anno.endLine!,
                          start_column: anno.startColumn,
                          end_column: anno.endColumn,
                          raw_details: anno.rendered,
                          message: anno.title!
                      }))
                  }
                : {
                      title: 'Clippy failed',
                      summary: `Running \`cargo clippy\` failed with exit code ${exitCode}`,
                      annotations: annotations.map((anno) => ({
                          annotation_level: anno.level === 'error' ? ('failure' as const) : ('warning' as const),
                          path: anno.file!,
                          start_line: anno.startLine!,
                          end_line: anno.endLine!,
                          start_column: anno.startColumn,
                          end_column: anno.endColumn,
                          raw_details: anno.rendered,
                          message: anno.title!
                      }))
                  }
    });

    info(`Clippy exited with code ${exitCode}`);
    process.exitCode = exitCode;
}

main().catch((ex) => {
    error(`Unable to run @augu/clippy-action:\n${ex}`);
    process.exit(1);
});
