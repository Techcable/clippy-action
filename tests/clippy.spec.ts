/*
 * üêª‚Äç‚ùÑÔ∏èüì¶ clippy-action: GitHub action to run Clippy, an up-to-date and modern version of actions-rs/clippy
 * Copyright 2023 Noel Towa <cutie@floofy.dev>
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { type Renderer, getClippyOutput, renderMessages } from '../src/clippy';
import { beforeEach, test, expect, afterAll } from 'vitest';
import type { AnnotationProperties } from '@actions/core';
import { mockProcessStdout } from 'vitest-mock-process';
import { dirname, resolve } from 'path';
import { fileURLToPath } from 'url';
import { which } from '@actions/io';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

const kRenderQueue: Record<'info' | 'warning' | 'error', [{ rendered: string; properties?: AnnotationProperties }][]> =
    {
        warning: [],
        error: [],
        info: []
    };

const kMockedRenderer: Renderer = {
    warning(message, properties) {
        kRenderQueue.warning.push([{ rendered: message, properties }]);
    },
    error(message, properties) {
        kRenderQueue.error.push([{ rendered: message, properties }]);
    },
    info(message) {
        kRenderQueue.info.push([{ rendered: message }]);
    }
};

const flushRenderQueue = () => {
    for (const prop of Object.keys(kRenderQueue)) {
        kRenderQueue[prop] = [];
    }
};

const stdoutMock = mockProcessStdout();

afterAll(() => {
    stdoutMock.mockReset();
});

beforeEach(() => {
    // Flush the renderer message queue so we get fresh instances
    // in each test.
    flushRenderQueue();
});

test('clippy error + warning', async () => {
    const cargoPath = await which('cargo', true);
    const [exitCode, pieces] = await getClippyOutput(
        {
            'working-directory': resolve(__dirname, '__fixtures__', 'unbased-codebase'),
            'all-features': false,
            forbid: [],
            allow: [],
            deny: [],
            warn: [],
            args: []
        },
        cargoPath
    );

    expect(exitCode).toBe(101 /* the code couldn't be compiled correctly */);
    renderMessages(pieces, kMockedRenderer);

    expect(kRenderQueue.warning).toMatchSnapshot();
    expect(kRenderQueue.error).toMatchSnapshot();
});

test('can we get zero warning and error messages in __fixtures__/no-clippy-error', async () => {
    const cargoPath = await which('cargo');
    const [exitCode, pieces] = await getClippyOutput(
        {
            'working-directory': resolve(__dirname, '__fixtures__', 'no-clippy-error'),
            'all-features': false,
            forbid: [],
            allow: [],
            deny: [],
            warn: [],
            args: []
        },
        cargoPath
    );

    expect(exitCode).toBe(0);
    renderMessages(pieces, kMockedRenderer);

    //expect(kRenderQueue).toStrictEqual({
    //    warning: [],
    //    error: [],
    //    info: []
    //});
});
