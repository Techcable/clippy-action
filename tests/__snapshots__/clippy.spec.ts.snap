// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`clippy error + warning 1`] = `
[
  [
    {
      "properties": {
        "endColumn": 14,
        "endLine": 4,
        "startColumn": 9,
        "startLine": 4,
        "title": "variable does not need to be mutable",
      },
      "rendered": "warning: variable does not need to be mutable
 --> src/main.rs:4:9
  |
4 |     let mut a = Box::new(1);
  |         ----^
  |         |
  |         help: remove this \`mut\`
  |
  = note: \`#[warn(unused_mut)]\` on by default

",
    },
  ],
]
`;

exports[`clippy error + warning 2`] = `
[
  [
    {
      "properties": {
        "endColumn": 24,
        "endLine": 5,
        "startColumn": 8,
        "startLine": 5,
        "title": "equal expressions as operands to \`==\`",
      },
      "rendered": "error: equal expressions as operands to \`==\`
 --> src/main.rs:5:8
  |
5 |     if *a + 1 == *a + 1 {}
  |        ^^^^^^^^^^^^^^^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#eq_op
  = note: \`#[deny(clippy::eq_op)]\` on by default

",
    },
  ],
]
`;
